src/main/java/groovy/lang/MetaClassImpl.java

```diff
@@ -2085,73 +2089,65 @@ public void setProperty(Object object, Object newValue) {
         checkInitalised();
 
         //----------------------------------------------------------------------
-        // turn getProperty on a Map to get on the Map itself
+        // getter
         //----------------------------------------------------------------------
-        if (!isStatic && this.isMap) {
-            return new MetaProperty(name, Object.class) {
-                @Override
-                public Object getProperty(Object object) {
-                    return ((Map) object).get(name);
-                }
-
-                @Override
-                public void setProperty(Object object, Object newValue) {
-                    throw new UnsupportedOperationException();
-                }
-            };
-        }
-
         Tuple2<MetaMethod, MetaProperty> methodAndProperty = createMetaMethodAndMetaProperty(sender, theClass, name, useSuper, isStatic);
         MetaMethod method = methodAndProperty.getV1();
 
-        //----------------------------------------------------------------------
-        // getter
-        //----------------------------------------------------------------------
-        MetaProperty mp = methodAndProperty.getV2();
+        if (method == null || "class".equals(name) || (isMap && "empty".equals(name))) {
```

Comment:
It's better to extract the common checking logic, here is same code at line 1981:
`if (method == null || "class".equals(name) || (isMap && "empty".equals(name)))`

Reply:
Why is it better to extract?  `getProperty` and `getEffectiveMetaProperty` are essentially the same general structure, only differing in a few details.  I don't really have the energy to extract the true repeated structure(s).  You can follow the two methods and see the parallel, so I think it is decent in this state.